### PR TITLE
gh-102840: Fix confused traceback when `floordiv` or `mod` operations happens between `Fraction` and `complex` objects

### PR DESCRIPTION
**Pull request by Eclips4**
* Issue: gh-102840

### CODE REVIEW

**Comment by Eclips4** [comment]
cc @skirpichev

File: Lib/fractions.py
Comment (skirpichev) [inline Lib/fractions.py]: Please check test coverage of this line.  I doubt there is a test - you could add one with example from the issue.  The module should have 100% test coverage (by Lib/test/test_fractions.py).

File: Lib/fractions.py
Comment (skirpichev) [inline Lib/fractions.py]: Ditto here.

**Review by skirpichev — CHANGES_REQUESTED** [review]

File: Lib/fractions.py
Comment (Eclips4) [reply Lib/fractions.py]: Thanks!

File: Lib/test/test_fractions.py
Comment (skirpichev) [inline Lib/test/test_fractions.py]:
```suggestion
        # See issue gh-102840 for more details.
```

Comment (skirpichev) [reply Lib/test/test_fractions.py]: Except for above nitpick, seem fine for me.
Personally, I would put these tests to the testMixedArithmetic(), but I think this is ok too.

**Review by skirpichev — CHANGES_REQUESTED** [review]

File: Lib/test/test_fractions.py
Comment (Eclips4) [reply Lib/test/test_fractions.py]: > Personally, I would put these tests to the testMixedArithmetic(), but I think this is ok too.

Good point. Let's see what the core developers will say.

**Comment by Eclips4** [comment]
Also, I think there no need in `NEWS` entry. But if someone want it - I'll add it.

**Review by skirpichev — APPROVED** [review]
LGTM

**Comment by arhadthedev** [comment]
@mdickinson (as a fraction expert)

File: Lib/test/test_fractions.py
Comment (Eclips4) [inline Lib/test/test_fractions.py]:
```suggestion
    def test_traceback_in_mod_and_floordiv_with_complex_objects(self):
```

File: Lib/test/test_fractions.py
Comment (serhiy-storchaka) [inline Lib/test/test_fractions.py]: Also test divmod.

File: Lib/test/test_fractions.py
Comment (serhiy-storchaka) [inline Lib/test/test_fractions.py]: You can use `operator`.

**Review by serhiy-storchaka — COMMENTED** [review]
I do not think that silencing TypeError for any operation is right. It can silence some unexpected errors, it is also not very efficient. It would be better to just not try to perform unsupported operation. I suggest to add an optional parameter in `_operator_fallbacks()` to specify whether try to handle complex numbers or not. We know which operations are not supported by `complex`.

**Comment by mdickinson** [comment]
> I suggest to add an optional parameter in `_operator_fallbacks()` to specify whether try to handle complex numbers or not.

+1. This sounds like the right approach to me.

**Comment by Eclips4** [comment]
Do we need `NEWS` entry here? It's improves the user experience, so in my opinion - yes.

File: Lib/fractions.py
Comment (serhiy-storchaka) [inline Lib/fractions.py]: Too long line.
```suggestion
    def _operator_fallbacks(monomorphic_operator, fallback_operator,
                            handle_complex=False):
```

File: Lib/test/test_fractions.py
Comment (serhiy-storchaka) [inline Lib/test/test_fractions.py]: `'` could be moved into `message` format string.

**Review by serhiy-storchaka — COMMENTED** [review]
This is a user visible change, so a NEWS entry will not make bad.

It will help if users wonder about confusing error messages in older Python.

File: Lib/fractions.py
Comment (skirpichev) [inline Lib/fractions.py]: Not sure if this is a big deal, but this value fits most cases, isn't?  Maybe this should be a default?

**Review by skirpichev — APPROVED** [review]
LGTM, modulo minor nitpick about default value.

File: Lib/fractions.py
Comment (Eclips4) [reply Lib/fractions.py]: Indeed. This will help us keep the diff minimal.

File: Misc/NEWS.d/next/Library/2024-02-10-15-24-20.gh-issue-102840.4mnDq1.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2024-02-10-15-24-20.gh-issue-102840.4mnDq1.rst]: The last nitpick.
```suggestion
between instances of :class:`fractions.Fraction` and :class:`complex`.

```

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**
